Make `coreclr_initialize` run cleanly for `corewasmrun`

AaronRobinsonMSFT opened the pull request:
- Separate call invokers between JIT compatible interpreter and interpreter with Portable EntryPoints.
- Native asserts now emit a managed stackwalk with IR offsets of interpreter frames

Current output of `corewasmrun` running on Chrome.

```
coreclr_initialize succeeded - retval: 0x00000000
onExit, code: 0
```

Fixes https://github.com/dotnet/runtime/issues/119001

AaronRobinsonMSFT committed:
Initial P/Invokes on WASM

Separate call invokers between JIT compatible interpreter
and interpreter with Portable EntryPoints.

AaronRobinsonMSFT commented:
/azp list

jkotas commented on src/coreclr/vm/wasm/calldescrworkerwasm.cpp:
```diff
@@ -13,7 +13,7 @@ extern "C" void STDCALL CallDescrWorkerInternal(CallDescrData * pCallDescrData)
     {
         GCX_PREEMP();
-        pMethod->PrepareInitialCode(CallerGCMode::Coop);
+        (void)pMethod->DoPrestub(NULL /* MethodTable */, CallerGCMode::Preemptive);
```
```suggestion
        (void)pMethod->DoPrestub(NULL /* MethodTable */, CallerGCMode::Coop);
```
I think this should be `CallerGCMode::Coop` (or the default `CallerGCMode::Unknown` that takes a slower path).

This argument represents that GC mode that the method is expected to be called in. It does not represent the current GC mode. `CallDescrWorkerInternal` is only ever used to invoke managed methods that are expected to be called in cooperative mode.

AaronRobinsonMSFT replied:
Okay. I recall adding this and the name is very misleading then. I must have misunderstood something when I added it. An enum called `CallerGCMode` seems rather clear that it is the current state of the caller.

jkotas replied:
It is the state that the original caller is in before any transitions. `DoPrestub` has preemptive mode contract, so it would not make sense to pass in mode of the immediate `DoPrestub` caller.

jkotas commented on src/coreclr/vm/wasm/helpers.cpp:
```diff
@@ -478,17 +479,66 @@ extern "C" int32_t mono_wasm_browser_entropy(uint8_t* buffer, int32_t bufferLeng
+    int32_t numArgs = pMetaSig->NumFixedArgs();
+    switch (numArgs)
+    {
+        case 1: return (LPVOID)&CallFuncVoidI32;
```
Check the (normalized) argument types as well to make sure that they match?

AaronRobinsonMSFT replied:
This was a bit of a slap dash solution. I can start on a more robust validation scheme if you like. I suppose asserting certain invariants is warranted though.

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2297,7 +2297,17 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT, CallerGCMode callerGCMo
+        void* ilStubInterpData = PortableEntryPoint::GetInterpreterData(pCode);
+        _ASSERTE(ilStubInterpData != NULL);
+        SetInterpreterCode((InterpByteCodeStart*)ilStubInterpData);
+#else // !FEATURE_PORTABLE_ENTRYPOINTS
+        // Only vararg P/Invoke use shared stubs, they need a precode to push the hidden argument.
+        if (IsVarArg())
+            GetOrCreatePrecode();
```
```suggestion
```
It may make more sense to add `pNMD->GetOrCreatePrecode()` call at   https://github.com/dotnet/runtime/blob/f2b00689e5ab854bdb133014fd13f937fe1a1f76/src/coreclr/vm/dllimport.cpp#L5653-L5656. It will save IsVarArg check that is not cheap, and keep the varargs special handling in fewer places.

(I expect that we are not going to support varargs with `FEATURE_PORTABLE_ENTRYPOINTS`.)

AaronRobinsonMSFT replied:
Good idea.

jkotas commented on src/coreclr/vm/wasm/helpers.cpp:
```diff
@@ -478,17 +479,66 @@ extern "C" int32_t mono_wasm_browser_entropy(uint8_t* buffer, int32_t bufferLeng
+{
+    STANDARD_VM_CONTRACT;
+    _ASSERTE(CheckPointer(pMetaSig));
+    _ASSERTE(pMetaSig->IsReturnTypeVoid());
```
We will also need to check whether this is managed or unmanaged signature since managed vs. unmanaged calling conventions are different for `FEATURE_PORTABLE_ENTRYPOINTS` (managed one has the extra hidden argument).

jkotas commented on src/coreclr/vm/precode_portable.hpp:
```diff
@@ -31,6 +32,17 @@ class PortableEntryPoint final
+
+    // Query methods for entry point state.
+    bool IsInitialized() const { return _pMD != nullptr; }
+    bool IsByteCodeCompiled() const { return IsInitialized() && _pInterpreterData != nullptr; }
```
```suggestion
    bool IsByteCodeCompiled() const
    {
        _ASSERTE(IsInitialized());
        return _pInterpreterData != nullptr;
    }
```
We should never see uninitialized entrypoints in circulation. These should be asserts. (We can also assert that the canary is good in IsIniitalized. Maybe a better name for the method would be `IsValid`.

jkotas commented on src/coreclr/vm/precode_portable.cpp:
```diff
@@ -29,6 +29,16 @@ void* PortableEntryPoint::GetActualCode(TADDR addr)
+    _ASSERTE(actualCode != NULL);
+
+    PortableEntryPoint* portableEntryPoint = ToPortableEntryPoint(addr);
+    _ASSERTE(portableEntryPoint->_pActualCode == NULL);
```
I do not expect this assert to hold. This is likely going to be updated in lock-free manner.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10871,12 +10867,22 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
+                LPVOID targetAddr;
+#ifdef FEATURE_PORTABLE_ENTRYPOINTS
+                accessType = IAT_VALUE;
+                targetAddr = (LPVOID)helperMD->GetPortableEntryPoint();
```
We also need to change how the `pNativeEntrypoint` gets returned. Otherwise, there would be an ambiguity about whether `IAT_VALUE` means portable entrypoint or native entrypoint (aka "actual code").

AaronRobinsonMSFT replied:
Done.

jkotas replied:
I do not think we want to do any bit flipping for entrypoints on wasm. As you have discovered, low order fit bit flipping would not work on wasm anyway since the native function pointers on wasm are ordinals starting with 1. I do not think that anything in wasm spec says that the high order bit of function pointer ordinal is unused.

I think we want to create and return pointer to PortableEntryPoint even for the native helpers here.

I am a bit surprised that we are able to go through whole `coreclr_initialize` with this issue. Are we lucky that we only ever call the managed JIT helpers during `coreclr_initialize`?

radekdoulik replied:
> I am a bit surprised that we are able to go through whole coreclr_initialize with this issue. Are we lucky that we only ever call the managed JIT helpers during coreclr_initialize?

Possibly, I checked the helper opcodes, not sure whether the JIT helpers are also called through different path.

During `coreclr_initialize` we call just a few. I checked the log output and currently the managed code involved in initialization could possibly call these
```
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_CHKCASTANY,               NULL,   METHOD__CASTHELPERS__CHKCASTANY)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_CHKCASTCLASS,             NULL,   METHOD__CASTHELPERS__CHKCASTCLASS)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_CHKCASTCLASS_SPECIAL,     NULL,   METHOD__CASTHELPERS__CHKCASTCLASSSPECIAL)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_GET_GCSTATIC_BASE,                JIT_GetGCStaticBase, METHOD__STATICSHELPERS__GET_GC_STATIC)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_INITCLASS,           NULL, METHOD__INITHELPERS__INITCLASS)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_ISINSTANCEOFCLASS,        NULL,   METHOD__CASTHELPERS__ISINSTANCEOFCLASS)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE,  NULL,   METHOD__RT_TYPE_HANDLE__GETRUNTIMETYPEFROMHANDLE)
inc/jithelpers.h:    DYNAMICJITHELPER(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE_MAYBENULL,    NULL,   METHOD__RT_TYPE_HANDLE__GETRUNTIMETYPEFROMHANDLEMAYBENULL)
```
which look like all managed except one.

`JIT_GetGCStaticBase` is possibly native, the `JIT_GetGCStaticBase` is defined as NULL though, so I think we use the managed version of it. And indeed, I see
```
Compiled method: .System.Runtime.CompilerServices.StaticsHelpers:GetGCStaticBase(ptr)
```
in the log.

The helper opcodes should be fine though as they use `GetPossiblyIndirectHelper`, which calls `PortableEntryPoint::GetActualCode` and that removes the mask (bit). Again, maybe we call the JIT helpers elsewehere too?

I agree that not using the mask would be cleaner, maybe add TODO-WASM comment for now?

BTW, the function indices start from 0.
```
> wa-info -vv ../../artifacts/obj/coreclr/browser.wasm.Debug/hosts/corewasmrun/corewasmrun.wasm
...
Reading section:    Custom size:      3023182 name: name
  module name: corewasmrun.wasm
      function names count: 33536
        function idx: 0 name: __assert_fail
        function idx: 1 name: abort
        function idx: 2 name: exit
...
```

jkotas replied:
This is broken in multiple ways - see my other feedback in https://github.com/dotnet/runtime/pull/119217#discussion_r2314120083 . I think we should do this properly so that we do not have to keep coming back to this.

AaronRobinsonMSFT committed:
Apply feedback.
Add simple interpreter managed frame stack walker when
native assert fires. This will help debugging asserts in interpreter
scenarios.

AaronRobinsonMSFT committed:
Narrow use of printing interpreter stack on assert to WASM.

AaronRobinsonMSFT committed:
Remvove white space edit.

AaronRobinsonMSFT committed:
Handle native helpers with portable entry points

AaronRobinsonMSFT committed:
Switch to high order bits for native entry points

AaronRobinsonMSFT committed:
Rework WASM thunk look-up

AaronRobinsonMSFT committed:
Fix Release build.

janvorli commented on src/coreclr/vm/precode_portable.hpp:
```diff
@@ -31,6 +34,32 @@ class PortableEntryPoint final
+        return _pActualCode != nullptr;
+    }
+
+    bool IsReadyForR2R() const
```
A nit - would it make sense naming this `IsReadyForAOT` or `IsReadyForAOTCall` instead?

janvorli commented on src/coreclr/vm/callstubgenerator.cpp:
```diff
@@ -1902,4 +1902,122 @@ CallStubGenerator::ReturnType CallStubGenerator::GetReturnType(ArgIterator *pArg
     return ReturnTypeVoid;
 }
 
+static CallStubHeader *UpdateCallStubForMethod(MethodDesc *pMD)
```
What was the reason for moving these methods to the callstubgenerator? I would prefer keeping only the implementation oof the CallStubGenerator class in this file.

janvorli replied:
Maybe `nativecodeinvocation.cpp`, similar to `reflectioninvocation.cpp` naming.

jkotas commented on src/coreclr/vm/precode_portable.hpp:
```diff
@@ -31,6 +34,32 @@ class PortableEntryPoint final
+    // Query methods for entry point state.
+    bool IsValid() const;
+
+    bool IsByteCodeCompiled() const
```
This name makes is sound like that the byte code was compiled into something that's not right. The byte code is the product of the compilation.

I would call this `HasInterpreterCode`. I think we use "Interpreter Code" in this context in majority of the codebase (e.g. `MethodDesc::GetInterpreterCode`).

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10871,12 +10867,22 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
 
```
The current scheme where `hlpFuncTable` entry that is less than `DYNAMIC_CORINFO_HELP_COUNT` implies dynamic helper won't work reliably for wasm. Wasm function pointers are ordinals that start with like `1`, so this trick won't work.

For wasm, I think we need to change `hlpFuncTable` that has just native code pointers, and separate table that has the information about dynamic entry points (or change `hlpFuncTable` to be array of structs).

AaronRobinsonMSFT commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1062,10 +1062,12 @@ void InitThreadManager()
     CONTRACTL_END;
 
+#ifndef TARGET_WASM
     // All patched helpers should fit into one page.
```
No, there are a bunch of uses of it. None of them make a lot of sense, but I'm not sure what to replace them with on WASM. All involve write barrier logic.

AaronRobinsonMSFT commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -224,9 +224,13 @@ void* DispatchCallSimple(
-    PORTABILITY_ASSERT("wasm need to fill call description data");
-#endif
+#ifdef FEATURE_PORTABLE_ENTRYPOINTS
+    callDescrData.pMD = PortableEntryPoint::GetMethodDesc(pTargetAddress);
```
Yep, done.

AaronRobinsonMSFT commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -224,9 +224,13 @@ void* DispatchCallSimple(
+    callDescrData.nArgsSize = numStackSlotsToCopy * sizeof(ARGHOLDER_TYPE);
+
+    TransitionBlock block{};
+    callDescrData.pTransitionBlock = &block;
```
Is this needed for WASM? The `TransitionBlock` on WASM has little if any content at present. Is this a temporary state?

jkotas replied:
I am not sure why it was needed on wasm. It should not be needed on wasm here to make wasm look more like other architectures.

AaronRobinsonMSFT replied:
> I am not sure why it was needed on wasm. I

Perfect. Thank you.

jkotas replied:
I would create a dummy in `CallDescrWorkerInternal` or in `ExecuteInterpretedMethodWithArgs` for now.

jkotas reviewed (commented):
CallerGCMode should be Coop.

AaronRobinsonMSFT commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -500,20 +433,25 @@ LONG IgnoreCppExceptionFilter(PEXCEPTION_POINTERS pExceptionInfo, PVOID pv)
-// Wrapper around MethodDesc::PrepareInitialCode to handle possible managed exceptions thrown by it.
-void PrepareInitialCode(MethodDesc *pMD)
+// Wrapper around MethodDesc::DoPrestub to handle possible managed exceptions thrown by it.
+static void CallPreStub(MethodDesc *pMD, CallerGCMode callerGCMode = CallerGCMode::Preemptive)
```
Yep.

AaronRobinsonMSFT committed:
Feedback.

AaronRobinsonMSFT commented on src/coreclr/vm/precode_portable.hpp:
```diff
@@ -22,7 +25,7 @@ class PortableEntryPoint final
 
 private:
-    void* _pActualCode;
+    VolatilePtr<BYTE> _pActualCode;
```
Yep, that works too.

AaronRobinsonMSFT committed:
Use Volatile<void*>.
Update ifdef targets.

jkotas commented on src/coreclr/vm/wasm/helpers.cpp:
```diff
@@ -478,17 +481,192 @@ extern "C" int32_t mono_wasm_browser_entropy(uint8_t* buffer, int32_t bufferLeng
+
+#undef ARG
+
+    void* RetVoidThunks[] =
```
These can be `const` arrays. It does not matter a whole lot since this is temporary code.

AaronRobinsonMSFT replied:
Yep. I can mark them as such.

AaronRobinsonMSFT committed:
Use nullptr when possible.
Mark static arrays as const.

kg commented on src/coreclr/vm/callingconvention.h:
```diff
@@ -1874,11 +1874,14 @@ int ArgIteratorTemplate<ARGITERATOR_BASE>::GetNextOffset()
 #elif defined(TARGET_WASM)
-    int cbArg = ALIGN_UP(StackElemSize(argSize), TARGET_POINTER_SIZE);
+    // See duplicate definition in interpreter code.
+#define INTERP_STACK_SLOT_SIZE 8
```
I'd like to see a comment in interpretershared mentioning that callingconvention.h needs to be updated if the define is changed

AaronRobinsonMSFT replied:
I've never found those comments very helpful. Instead I've used the exact same define name and indicated it is a duplicate. The first steps for these things are typically, "where is this used?", which should bring people here.

AaronRobinsonMSFT replied:
I was able to refactor this and avoid redeclaring the define.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -11,14 +11,36 @@
+
+    {
+        GCX_PREEMP();
+        InvokeManagedMethod(targetMethod, stack + callArgsOffset, stack + returnOffset, callTarget);
```
Why is this function called InvokeUnmanagedMethodWithTransition if it's calling InvokeManagedMethod?

AaronRobinsonMSFT replied:
Because the current `InvokeManagedMethod`, previously `InvokeCompiledMethod` does double duty in the interpreter path. @jkotas made this suggestion offline.

jkotas replied:
The current code makes assumption that the managed calling convention is same as the unmanaged calling convention. That's not a valid assumption. 

I expect that this will need to be redone to handle unmanaged calling conventions properly that you have touched upon on the teams chat. Once it is redone InvokeUnmanagedMwethod won't be calling InvokeManagedMethod anymore.

It may be a good idea to add `WASM-TODO: Handle unmanaged calling conventions` next to this to avoid confusion in the meantime.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -101,19 +128,20 @@ void InvokeDelegateInvokeMethod(MethodDesc *pMDDelegateInvoke, int8_t *pArgs, in
 }
 
-void InvokeCalliStub(PCODE ftn, CallStubHeader *stubHeaderTemplate, int8_t *pArgs, int8_t *pRet)
+void InvokeCalliStub(PCODE ftn, void *cookie, int8_t *pArgs, int8_t *pRet)
```
Why are we stripping the type information from this argument and then casting it back to CallStubHeader?

AaronRobinsonMSFT replied:
Beacuse this API is used in an agnostic manner. For the `CallStubGenerator` scenario it is a `CallStubHeader`, when `PortableEntryPoint`s are used it is currently a staticly defined thunk. See the other implementation in this PR's `src/coreclr/vm/wasm/helpers.cpp`.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -155,35 +191,50 @@ CallStubHeader *CreateNativeToInterpreterCallStub(InterpMethod* pInterpMethod)
+    int32_t frameCount = 0;
 
+    // Get the "capital F" frame and start walking.
+    for (Frame* pFrame = pThread->GetFrame(); pFrame != FRAME_TOP; pFrame = pFrame->PtrNextFrame())
```
Is it possible we enter DBG_PrintInterpreterStack with no frame(s) on the thread? I don't know the answer to this, just wondering since you put a call to this in the assert handler iirc

kg replied:
Also is it possible we have no pThread yet if we assert on a native thread that hasn't had a managed thread object set up yet, and then ->GetFrame crashes?

AaronRobinsonMSFT replied:
Right, the guard for no Thread is the issue. No capital F frames is already handled by `pFrame != FRAME_TOP`.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1835,9 +1890,25 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
 
+                    MethodDesc *pILTargetMethod = NULL;
+                    HELPER_FTN_P_P helperFtn = GetPossiblyIndirectHelper<HELPER_FTN_P_P>(pMethod, ip[2], &pILTargetMethod);
+                    if (pILTargetMethod != NULL)
+                    {
+                        returnOffset = ip[1];
+                        int stackOffset = pMethod->allocaSize;
+                        callArgsOffset = stackOffset;
+
+                        // Pass argument to the target method
+                        LOCAL_VAR(stackOffset, void*) = helperArg;
+
+                        targetMethod = pILTargetMethod;
+                        ip += 4;
+                        goto CALL_INTERP_METHOD;
+                    }
```
It would be nice if we could encapsulate this repeated logic somehow since it's fairly nuanced but I don't see an easy way to do it since we're populating locals and doing a goto...

radekdoulik replied:
We discussed that with JanV before and he didn't like the macros either, because they are difficult to debug.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1890,8 +1992,24 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
 
+                        // Pass arguments to the target method
+                        LOCAL_VAR(stackOffset, void*) = helperArg;
+                        LOCAL_VAR(stackOffset + INTERP_STACK_SLOT_SIZE, void*) = LOCAL_VAR(ip[3], void*);
```
Could `ip[3] == stackOffset`?

AaronRobinsonMSFT commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1855,9 +1926,26 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+
+                        // Pass arguments to the target method
+                        LOCAL_VAR(stackOffset, void*) = helperArg;
+                        LOCAL_VAR(stackOffset + INTERP_STACK_SLOT_SIZE, void*) = LOCAL_VAR(ip[2], void*);
```
Great question. I don't think it hurts to be defensive in this case. Will do the same below.

AaronRobinsonMSFT committed:
Protect against null thread.
Proactively record locals to avoid them being lost.

kg commented on src/coreclr/vm/wasm/cgencpu.h:
```diff
@@ -22,8 +22,8 @@
 {
-    _ASSERTE("The function is not implemented on wasm");
-    return 0;
+    const unsigned stackSlotSize = sizeof(void*);
```
Should this be the interpreter stack slot size? I'm not clear on what contract StackElemSize needs to fulfill.
According to GCC's docs, alloca does alignment:
> The __builtin_alloca function must be called at block scope. The function allocates an object size bytes large on the stack of the calling function. The object is aligned on the default stack alignment boundary for the target determined by the __BIGGEST_ALIGNMENT__ macro

And it looks like emscripten also uses clang's `__builtin_alloca`, which probably conforms to the same contract (I couldn't find LLVM's docs for it in a quick search)
So maybe we want to use `__BIGGEST_ALIGNMENT__` here because that's the alignment alloca tries to maintain for its stack?

Pointer size feels fine though, just wanted to sense whether this is 'correct' vs a temporary hack. My only concern is that if parts of the wasm/emscripten stack are aligning to a hard-coded 4 bytes, this would break on wasm64 in a mysterious way.

AaronRobinsonMSFT replied:
> Should this be the interpreter stack slot size?

Quite possibly. I selected `void*` since I was trying to use what would make sense for wasm64 and ideally it would "naturally" work. It was also before I discovered `INTERP_STACK_SLOT_SIZE`, so I'm fine using that here if that is more appropriate, which it seems like it would be. I defer to you on the clearest value here.

AaronRobinsonMSFT commented on src/coreclr/vm/precode_portable.cpp:
```diff
@@ -7,28 +7,53 @@
 #include "precode_portable.hpp"
 
+// WASM uses function indices so we should avoid using
+// low order bits for masking.
```
Yep, I'm removing a bunch of thing based on other feedback.

AaronRobinsonMSFT committed:
Feedback on alignment define values.

AaronRobinsonMSFT committed:
Fix up assert for marshalling checks on stack slot size.

AaronRobinsonMSFT commented on src/coreclr/vm/precode_portable.cpp:
```diff
@@ -75,6 +102,13 @@ void PortableEntryPoint::Init(MethodDesc* pMD)
+bool PortableEntryPoint::IsValid() const
+{
+    LIMITED_METHOD_CONTRACT;
+    _ASSERTE(_canary == CANARY_VALUE);
```
> I wonder whether IsValid should be entirely debug-only.

I'll go down this route. We tend not to have magic numbers in release builds.

AaronRobinsonMSFT committed:
Remove bit marking for native implemented helpers.
Add PCODE storage array for allocated helper wrappers.

AaronRobinsonMSFT committed:
Add casts to avoid build breaks.

AaronRobinsonMSFT committed:
Add TADDR cast.

AaronRobinsonMSFT committed:
Add more casts

AaronRobinsonMSFT committed:
Remove invalid cast.

jkotas commented on src/coreclr/vm/mlinfo.cpp:
```diff
@@ -2544,7 +2544,13 @@ void MarshalInfo::SetupArgumentSizes()
+    // Therefore, we only need assert it is sufficiently large.
+    _ASSERTE(targetPointerSize <= StackElemSize(TARGET_POINTER_SIZE, pointerIsValueType, pointerIsFloatHfa));
+#else
     _ASSERTE(targetPointerSize == StackElemSize(TARGET_POINTER_SIZE, pointerIsValueType, pointerIsFloatHfa));
```
I am not sure why this assert is useful. How is it relevant to the code that follows?

I think it would be better to delete it instead of ifdefing it out for wasm.

jkotas commented on src/coreclr/vm/wasm/cgencpu.h:
```diff
@@ -22,8 +23,8 @@
-    _ASSERTE("The function is not implemented on wasm");
-    return 0;
+    const unsigned stackSlotSize = INTERP_STACK_SLOT_SIZE;
+    return ALIGN_UP(parmSize, stackSlotSize);
```
StackElemSize is used to compute the amount of memory used to pass argument in native ABI. It does not make sense to me to tie it together with interpreter. I think it would be better to keep it as is with assert and add just do a local fix in callingconvention.h that's with the interpreter calling convention currently.

jkotas commented on src/coreclr/vm/callingconvention.h:
```diff
@@ -1874,11 +1874,11 @@ int ArgIteratorTemplate<ARGITERATOR_BASE>::GetNextOffset()
     return argOfs;
 #elif defined(TARGET_WASM)
-    int cbArg = ALIGN_UP(StackElemSize(argSize), TARGET_POINTER_SIZE);
+    int cbArg = ALIGN_UP(StackElemSize(argSize), INTERP_STACK_SLOT_SIZE);
```
```suggestion
    int cbArg = ALIGN_UP(argSize, INTERP_STACK_SLOT_SIZE);
```

AaronRobinsonMSFT committed:
Feedback

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -2544,6 +2551,30 @@ static const BinderMethodID hlpDynamicToBinderMap[DYNAMIC_CORINFO_HELP_COUNT] =
 #include "jithelpers.h"
 };
 
+bool VMHELPDEF::IsDynamicHelper(size_t& dynamicFtnNum) const
```
```suggestion
bool VMHELPDEF::IsDynamicHelper(DynamicCorInfoHelpFunc& dynamicFtnNum) const
```
Also, should this be `DynamicCorInfoHelpFunc*` to make it clear that this is out arg? It is unusual to return out args via `&` in this codebase.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10755,25 +10751,28 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
+    TADDR pfnHelper = helperDef.pfnHelper;
+
+    size_t dynamicFtnNum;
+    bool isDynamicMethod = helperDef.IsDynamicHelper(dynamicFtnNum);
+    if (isDynamicMethod)
```
```suggestion
    bool isDynamicHelper = helperDef.IsDynamicHelper(dynamicFtnNum);
    if (isDynamicHelper)
```
DynamicMethod is a thing that has nothing to do with this.

AaronRobinsonMSFT replied:
Oops. I thought I replaced all those.

jkotas commented on src/coreclr/vm/precode_portable.cpp:
```diff
@@ -8,24 +8,24 @@
+#define CANARY_VALUE 0x12345678
 #endif // HOST_64BIT
 
 bool PortableEntryPoint::IsNativeEntryPoint(TADDR addr)
```
```suggestion
bool PortableEntryPoint::HasNativeEntryPoint(TADDR addr)
```
This can return true for interpreted methods with native call stub, so `Has` makes more sense.

AaronRobinsonMSFT replied:
Good point.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10905,18 +10919,18 @@ PCODE CEECodeGenInfo::getHelperFtnStatic(CorInfoHelpFunc ftnNum)
-    _ASSERTE(pfnHelper != NULL);
+    _ASSERTE(pfnHelper != (TADDR)NULL);
 
     return GetEEFuncEntryPoint(pfnHelper);
```
```suggestion
    return pfnHelper;
```

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10859,41 +10846,68 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
+        TADDR entryPoint = GetEEFuncEntryPoint(pfnHelper);
+
+#ifdef FEATURE_PORTABLE_ENTRYPOINTS
+        NewHolder<PortableEntryPoint> portableEntryPoint = new PortableEntryPoint();
```
```suggestion
        PortableEntryPoint* portableEntryPoint = (void*)(SystemDomain::GetGlobalLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(sizeof(PortableEntryPoint));
```
Loader allocator may make more sense for this.

AaronRobinsonMSFT replied:
Backout for that is harder. That is why I went with `new`, where the backout logic is far simpler.

jkotas replied:
You can use `AllocMemHolder` that is about as hard to backout as NewHolder.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10859,41 +10846,68 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
     {
-        pNativeEntrypoint->accessType = IAT_VALUE;
-        pNativeEntrypoint->addr = (LPVOID)GetEEFuncEntryPoint(pfnHelper);
+        TADDR entryPoint = GetEEFuncEntryPoint(pfnHelper);
```
```suggestion
        TADDR entryPoint = pfnHelper;
```
`GetEEFuncEntryPoint` macro does make sense here. It used to extract actual code on platforms where function pointers did not point to actual code (we do not have any platform like that currently , and it is also used to DACize pointers to runtime functions (this file is not DAcized so it does not apply here).

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10859,41 +10846,68 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
         }
 
-        pfnHelper = LoadDynamicJitHelper((DynamicCorInfoHelpFunc)dynamicFtnNum).pfnHelper;
+        pfnHelper = LoadDynamicJitHelper((DynamicCorInfoHelpFunc)dynamicFtnNum);
```
With portable entrypoints, this returns `PortableEntryPoint*` but then we will go and wrap it in PortableEntryPoint a few lines later again since ` if (targetAddr == NULL)` condition is going to be true. I do not see how this can work.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10905,15 +10919,15 @@ PCODE CEECodeGenInfo::getHelperFtnStatic(CorInfoHelpFunc ftnNum)
+        pfnHelper = LoadDynamicJitHelper((DynamicCorInfoHelpFunc)dynamicFtnNum);
     }
 
     _ASSERTE(pfnHelper != NULL);
```
This is only used for R2R fixups. Once we get to then, I think we will require all JIT helpers to be precompiled so that we can call them directly without the PortableEnrypoint indirection. I do not think we will need the wrapping here. Chances are this method is going to be unreachable for wasm. 

I would leave this as is for now.

AaronRobinsonMSFT committed:
Feedback.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10859,41 +10845,70 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
         }
 
-        pfnHelper = LoadDynamicJitHelper((DynamicCorInfoHelpFunc)dynamicFtnNum).pfnHelper;
+        pfnHelper = LoadDynamicJitHelper(dynamicFtnNum);
```
This still does not look right. LoadDynamicJitHelper can return IL-based helper or dynamically assigned native helper. The dynamically assigned helper still needs wrapping by a portable entrypoint. The helpers backed by managed methods are the only ones where we can skip the wrapping.

jkotas replied:
I think that the maze of the different ifdefs is impossible to deal with well using interleaved ifdefs. I have pushed a commit into your PR with what I think it should look like. (I have not tested - please beat it into shape.)

I do not really like that `LoadDynamicJitHelper` can return either portable entrypoint or actual code. Refactoring that contract can be done separately.

jkotas committed:
Fix getHelperFtn for portable entrypoints

jkotas reviewed (approved):
LGTM (modulo fixes for breaks that I have introduced by the last commit :)

AaronRobinsonMSFT committed:
Nits

AaronRobinsonMSFT committed:
Make field more obvious.

AaronRobinsonMSFT committed:
Fix bug in casting.
